homeassistant/helpers/llm.py

```diff
@@ -181,14 +181,48 @@ def __init__(
         self.description = (
             intent_handler.description or f"Execute Home Assistant {self.name} intent"
         )
-        if slot_schema := intent_handler.slot_schema:
-            self.parameters = vol.Schema(slot_schema)
+        self.extra_slots = None
+        if not (slot_schema := intent_handler.slot_schema):
+            return
+
+        slot_schema = {**slot_schema}
+        extra_slots = set()
+
+        for field in ("preferred_area_id", "preferred_floor_id"):
+            if field in slot_schema:
+                extra_slots.add(field)
+                del slot_schema[field]
+
+        self.parameters = vol.Schema(slot_schema)
+        if extra_slots:
+            self.extra_slots = extra_slots
 
     async def async_call(
         self, hass: HomeAssistant, tool_input: ToolInput, llm_context: LLMContext
     ) -> JsonObjectType:
         """Handle the intent."""
         slots = {key: {"value": val} for key, val in tool_input.tool_args.items()}
+
+        area: ar.AreaEntry | None = None
+        floor: fr.FloorEntry | None = None
+        if self.extra_slots and llm_context.device_id:
+            device_reg = dr.async_get(hass)
+            device = device_reg.async_get(llm_context.device_id)
+
+            if device:
```

Comment:
Lets reduce scope of area and floor

```suggestion
        if self.extra_slots and llm_context.device_id:
            device_reg = dr.async_get(hass)
            device = device_reg.async_get(llm_context.device_id)

            area: ar.AreaEntry | None = None
            floor: fr.FloorEntry | None = None
            if device:
```